diff --git a/android/src/main/java/expo/modules/filesystem/FileSystemFileHandle.kt b/android/src/main/java/expo/modules/filesystem/FileSystemFileHandle.kt
index a39dfeb96731d8c75afa084742573ae0cebb0bc7..c66febfaa66ed222081a65e7608a638bc02b9448 100644
--- a/android/src/main/java/expo/modules/filesystem/FileSystemFileHandle.kt
+++ b/android/src/main/java/expo/modules/filesystem/FileSystemFileHandle.kt
@@ -2,6 +2,7 @@ package expo.modules.filesystem
 
 import android.content.ContentResolver
 import android.net.Uri
+import android.os.ParcelFileDescriptor
 import expo.modules.kotlin.exception.Exceptions
 import expo.modules.kotlin.sharedobjects.SharedRef
 import expo.modules.kotlin.types.Enumerable
@@ -51,7 +52,8 @@ enum class FileMode(val descriptor: String) : Enumerable {
 
 class FileSystemFileHandle private constructor(
   channel: FileChannel,
-  private val mode: FileMode
+  private val mode: FileMode,
+  private val parcelFileDescriptor: ParcelFileDescriptor? = null
 ) : SharedRef<FileChannel>(channel), AutoCloseable {
   companion object {
     fun forJavaFile(file: File, mode: FileMode): FileSystemFileHandle {
@@ -81,7 +83,7 @@ class FileSystemFileHandle private constructor(
         else -> throw Exceptions.IllegalArgument("Unsupported file mode: '$mode'")
       }
 
-      return FileSystemFileHandle(channel, mode)
+      return FileSystemFileHandle(channel, mode, pfd)
     }
   }
 
@@ -98,7 +100,11 @@ class FileSystemFileHandle private constructor(
   }
 
   override fun close() {
-    fileChannel.close()
+    try {
+      fileChannel.close()
+    } finally {
+      parcelFileDescriptor?.close()
+    }
   }
 
   fun read(length: Long): ByteArray {
